lib/phoenix/pubsub/redis_server.ex

```diff
@@ -0,0 +1,164 @@
+defmodule Phoenix.PubSub.RedisServer do
+  use GenServer
+  require Logger
+  alias Phoenix.PubSub.RedisServer
+
+  @moduledoc """
+  The server for the RedisAdapter
+
+  See `Phoenix.PubSub.RedisAdapter` for details and configuration options.
+  """
+
+  @derive [Access]
+  defstruct eredis_sub_pid: nil,
+            status: :disconnected,
+            reconnect_attemps: 0,
+            node_ref: nil,
+            local_pid: nil,
+            opts: []
+
+  @defaults [host: "127.0.0.1", port: 6379, password: ""]
+
+  @max_connect_attemps 3   # 15s to establish connection
+  @reconnect_after_ms 5000
+
+  @doc """
+  Starts the server
+
+  TODO document options
+  """
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts, name: Dict.fetch!(opts, :name))
+  end
+
+  @doc """
+  Initializes the server.
+
+  An initial connection establishment loop is entered. Once `:eredis_sub`
+  is started successfully, it handles reconnections automatically, so we
+  pass off reconnection handling once we find an initial connection.
+  """
+  def init(opts) do
+    {:ok, local_pid} = Phoenix.PubSub.Local.start_link()
+    opts = Dict.merge(@defaults, opts)
+    opts = Dict.merge(opts, host: String.to_char_list(to_string(opts[:host])),
+                            password: String.to_char_list(to_string(opts[:password])))
+
+    Process.flag(:trap_exit, true)
+    send(self, :establish_conn)
+
+    {:ok, struct(RedisServer, opts: opts,
```

Comment:
I meant the `struct/2` call instead of `%RedisServer{...}`. :D But getting rid of the struct also works (:+1: from me).